OpenTrons/opentrons-api
# fix(api,update-server): Store names safely and reject names that would be invalid

# Overview

This PR fixes several robot-side bugs related to robot names that contain more than just simple alphanumeric characters.

Closes #10197. Closes #10413. Closes #10687.


# Changelog

* Reject invalid strings for the robot's name.
    * A string is "invalid" if Avahi would reject it as the DNS-SD instance name, or if `hostnamectl set-hostname --pretty` would reject it as the pretty hostname.
    * We implement this validation logic ourselves, which kind of sucks.
    * If I did this right, this effectively fixes #10687 by, hopefully, never feeding Avahi a name that it won't like.
* **POTENTIALLY BREAKING CHANGE:** When an HTTP client tries to set an invalid name via `POST /server/name`, return a 400 error if the name is invalid.
* When rewriting `/etc/machine-info` to store a new pretty hostname, use quoting and backslash-escaping to preserve special characters.
    * We have to implement this escaping logic ourselves to follow `/etc/machine-info`'s nonstandard rules, unfortunately.
    * This fixes #10197.
* Restart `systemd-hostnamed` after we rewrite `/etc/machine-info` so it immediately picks up our modifications.
    * I'm pretty sure this fixes #10413.
* Make a small tweak to how *`api`* retrieves the pretty hostname, to keep it in sync with `update-server`.

# Review requests



# Manual testing

You'll need a real OT-2 to test this. Do a top-level `make push` to push both the `update-server` and `api` packages to it.

* Make sure setting the robot's name through the Opentrons App still works.
* Try manually changing the robot's name with HTTP `POST /server/name {"name": "new name"}`.
    * Try using crazy strings: empty, very long, special characters, beginning and ending with whitespace, etc.
    * The endpoint should either return 200 and set the name to what you specified, or return 400 and leave the name unchanged.
    * The endpoint should never return 500.
* Try `GET /server/name`, `GET /server/update/health`, and `GET /health` to see that the name is kept in sync between all 3 HTTP endpoints where it shows up.
     * You can also try `cd discovery-client && yarn run discovery` and look for the robot's 3 `name` fields to test the same thing.
* Try `dns-sd -B` to make sure that the robot's advertised instance name is always exactly what you set.

# Risk assessment

# Other approaches considered

* **Give up on nice names.** Disallow everything that would require quoting or escaping; allow only ASCII A-Z, a-z, and 0-9.
    * This thought makes me existentially sad, so I don't like it.
* **Shell out** to `hostnamectl set-hostname --pretty` instead of manually writing `/etc/machine-info`.
    * I tried a few things, but I couldn't get this to work. See https://github.com/Opentrons/opentrons/pull/10219#discussion_r869492163.
* **Totally decouple our code from the systemd pretty hostname.** Persist the robot's name in our own file somewhere, in a format that's easier to deal with.
    * This might be the right move, but it felt to me like questioning too much for this PR.
* **Avoid implementing our own validation logic.** Instead, handle errors from systemd and Avahi gracefully, and use them as the single source of truth for what's valid and invalid. [Ask for forgiveness instead of permission](https://docs.python.org/3.7/glossary.html#term-eafp), in other words.
    * I'm pretty conflicted about this.
    * On one hand, it would be really nice if we could avoid reimplementing the validation logic ourselves.  
    * On the other hand, we then have to implement *transactional recovery* logic. 
      ```mermaid
      graph TD
          set_avahi[Set Avahi service name] --> valid_for_avahi{Did it work?}
          valid_for_avahi -->|Yes| set_pretty_hostname[Set pretty hostname]
          valid_for_avahi -->|No| restore_avahi[Restore prior Avahi service name]
          set_pretty_hostname --> valid_as_pretty_hostname{Did it work?}
          valid_as_pretty_hostname -->|Yes| Done
          valid_as_pretty_hostname -->|No| restore_avahi_2[Restore prior Avahi service name] --> restore_pretty[Restore prior pretty hostname]
      ```
      This seemed doable, but messy in its own ways. Overall, it felt like a wash to me, versus the "look before you leap" approach that this PR currently takes.
